### ISSUE TITLE
Create network before accessing kube-api

### ISSUE THREAD

**Issue by Mik4sa** [flannel-io/flannel#1772]
## Expected Behavior
Flannel should create the network before accessing/querying kube-api (the api server). The network is needed in some cases before the api server can be reached.

## Current Behavior
Currently kube-api is accessed for multiple things before the network is created.

## Possible Solution
I don't really know the code of flannel but would it be possible to just change the order of this calls?
For the case of the node name (which is received from kube-api with the help of the pod name and namespace) the node name could be set directly into the pod from the manifest.

## Context
You better follow this issue https://github.com/kubernetes-sigs/sig-windows-tools/issues/277
There is the whole context.

In short: I would like to use the in cluster config for flannel on windows but this doesn't work because the in cluster config accesses kube-api over a kubernetes service which isn't reachable at this moment because the hns network isn't created at this moment.

## Your Environment

* Flannel version: 0.21.5
* Backend used (e.g. vxlan or udp): vxlan
* Etcd version:
* Kubernetes version (if used): 1.27.1
* Operating System and version: Windows Server 2022 Standard, Version 21H2 (Build 20348.1726)

**Comment by Mik4sa** [comment]
So I looked into the code of v0.21.5 and this should indeed work. Some refactoring is required though. Basically the connection is only need to store annotations at the node. Currently they are set just before the network is created. I decoupled that on a test branch so that the network is created on beforehand. After the network is created I also started kube-proxy which then found the network and did it's thing. After that I was able to reach the api server through the kubernetes service. Finally I was able to store that annotations.

Honestly I'm unsure whether I can come up with a good solution. I don't really know go or flannel. I guess that someone else should do this.

**Comment by manuelbuil** [comment]
Kube-proxy runs in hostNetwork mode, meaning that it uses the root network namespace and thus it can reach kube-api. Then it sets the DNAT 10.43.0.1:443 => NODE_IP:6443. Flannel is also running in hostNetwork mode: https://github.com/flannel-io/flannel/blob/master/Documentation/kube-flannel.yml#L134 and thus, it can reach `NODE_IP:6443` and thus query kube-api. Then, it creates the network infrastructure for pods that are not in hostNetwork mode.

Unfortunately, I am not familiar with how Kubernetes works in Windows and I thought it was exactly the same. Do you know if there is any difference with what I described above?

**Comment by Mik4sa** [comment]
Neither me, maybe @jsturtevant or @marosset?

**Comment by jsturtevant** [comment]
Yes, it does run on the hostnetwork with Hostprocess Containers.  There must be something else going on that is blocking access to the api server.

**Comment by Mik4sa** [comment]
The behavior I described also matches for the windows node itself (the host). So outside of any container network etc
I can try to check the network flow. Maybe there is something I can spot. Unfortunately I no longer have time for this during this week

For reference. These are two azure VM's inside the same virtual network. Both have the default network security groups ("type" basic)

**Comment by Mik4sa** [comment]
@manuelbuil So looks like kube-proxy for windows itself also needs the hns network before it can run, see here:
https://github.com/kubernetes/kubernetes/blob/7f6f68fdabc4df88cfea2dcf9a19b2b830f1e647/pkg/proxy/winkernel/proxier.go#L217-L225:
> ```golang
> func getNetworkInfo(hns HostNetworkService, hnsNetworkName string) (*hnsNetworkInfo, error) {
> 	hnsNetworkInfo, err := hns.getNetworkByName(hnsNetworkName)
> 	for err != nil {
> 		klog.ErrorS(err, "Unable to find HNS Network specified, please check network name and CNI deployment", "hnsNetworkName", hnsNetworkName)
> 		time.Sleep(1 * time.Second)
> 		hnsNetworkInfo, err = hns.getNetworkByName(hnsNetworkName)
> 	}
> 	return hnsNetworkInfo, err
> }
> ```

So for me it looks like we cant solve this in kube-proxy and we would need to make a change in flannel. So we are back to my original description.
Do you understand/agree?

**Comment by manuelbuil** [comment]
@Mik4sa thanks for the investigation. In Linux, kube-proxy rules are generally implemented in iptables. iptables are decoupled from the network resources created by the CNI plugin. By reading your description, it seems this is not the case for Windows, and kube-proxy rules are applied to the network resources created by the CNI plugin (hns network). Therefore, your solution moves the creation of the network to an earlier point, right? So that kube-proxy can already insert its rules. As long as that does not break things, I'm fine with that. Could you create a PR with the changes? We can discuss further over there

**Comment by Mik4sa** [comment]
> By reading your description, it seems this is not the case for Windows, and kube-proxy rules are applied to the network resources created by the CNI plugin (hns network).

Good summary. Yes this is how I understood everything so far.

> Therefore, your solution moves the creation of the network to an earlier point, right? So that kube-proxy can already insert its rules.

This would be the case, yes.

> Could you create a PR with the changes?

I'm not really familiar with go so I'm quite unsure whether I would be able to accomplish this. But anyway, I just found an alternative solution for my problem yesterday. Instead of talking to the api server over the Kubernetes service (which needs kube-proxy to be running) I now talk to it directly by setting the env variables appropriately.

I'm going to wait for the approval of that possible solution. If it gets accepted I'm going to close this issue since no longer needed

**Comment by manuelbuil** [comment]
> > By reading your description, it seems this is not the case for Windows, and kube-proxy rules are applied to the network resources created by the CNI plugin (hns network).
> 
> Good summary. Yes this is how I understood everything so far.
> 
> > Therefore, your solution moves the creation of the network to an earlier point, right? So that kube-proxy can already insert its rules.
> 
> This would be the case, yes.
> 
> > Could you create a PR with the changes?
> 
> I'm not really familiar with go so I'm quite unsure whether I would be able to accomplish this. But anyway, I just found an alternative solution for my problem yesterday. Instead of talking to the api server over the Kubernetes service (which needs kube-proxy to be running) I now talk to it directly by setting the env variables appropriately.
> 

Up to you. But don't be afraid, we can help you in the review :). I don't think anyone here is a go expert, so we are all learning by doing 

> I'm going to wait for the approval of that possible solution. If it gets accepted I'm going to close this issue since no longer needed

That solution would work as well. The disadvantage is that you are creating a SPF (Single Point of Failure). If that server goes down, flannel won't be able to reach kube-api. If you deploy on HA mode (e.g. by running 3 servers) and use the kubernetes service instead, flannel would be able to connect to three different endpoints and thus you remove that SPF.

**Comment by Mik4sa** [comment]
> That solution would work as well. The disadvantage is that you are creating a SPF (Single Point of Failure). If that server goes down, flannel won't be able to reach kube-api. If you deploy on HA mode (e.g. by running 3 servers) and use the kubernetes service instead, flannel would be able to connect to three different endpoints and thus you remove that SPF.

Actually the current solution works the same but has some dependencies to the kube-system namespace. Which is what I decoupled. So I didn't make it more worse 😅

But good point though. One more question. Would the SPF be only for initialization of flannel or for the whole lifetime of my cluster?
If first, I would be fine with the SPF. Since this should be a small risk.
But if it's second we should (somehow) use the Kubernetes service 🤔. But then in another issue down in sig-windowstools

**Comment by manuelbuil** [comment]
> > That solution would work as well. The disadvantage is that you are creating a SPF (Single Point of Failure). If that server goes down, flannel won't be able to reach kube-api. If you deploy on HA mode (e.g. by running 3 servers) and use the kubernetes service instead, flannel would be able to connect to three different endpoints and thus you remove that SPF.
> 
> Actually the current solution works the same but has some dependencies to the kube-system namespace. Which is what I decoupled. So I didn't make it more worse sweat_smile
> 
> But good point though. One more question. Would the SPF be only for initialization of flannel or for the whole lifetime of my cluster? If first, I would be fine with the SPF. Since this should be a small risk. But if it's second we should (somehow) use the Kubernetes service thinking. But then in another issue down in sig-windowstools

I am not sure how you made flannel contact kube-api but based on my assumption, I guess for the whole lifetime because flannel needs to check kube-api for new nodes
